core/src/main/java/org/elasticsearch/cluster/ClusterName.java

```diff
@@ -33,6 +33,8 @@
     public static final Setting<ClusterName> CLUSTER_NAME_SETTING = new Setting<>("cluster.name", "elasticsearch", (s) -> {
         if (s.isEmpty()) {
             throw new IllegalArgumentException("[cluster.name] must not be empty");
+        } else if(s.contains(":")){
+            throw new IllegalArgumentException("[cluster.name] must not contains ':' character");
```

Comment:
Also, I think that the exception message should include the offending cluster the name. The pattern that we usually use is 

```
[cluster.name] must not contain ':' but was [<offending cluster name>]
```

This is another argument against introducing indirection here.